stream_muxer/mplex/mplex_stream.py

```diff
@@ -40,50 +40,57 @@ def get_flag(self, action):
         read messages associated with stream from buffer til end of file
         :return: bytes of input
         """
-        return await self.muxed_conn.read_buffer(self.stream_id)
+        return await self.mplex_conn.read_buffer(self.stream_id)
 
     async def write(self, data):
         """
         write to stream
         :return: number of bytes written
         """
-        return await self.muxed_conn.send_message(self.get_flag("MESSAGE"), data, self.stream_id)
+        return await self.mplex_conn.send_message(self.get_flag("MESSAGE"), data, self.stream_id)
 
     async def close(self):
         """
-        close stream
+        Closing a stream closes it for writing and closes the remote end for reading
+        but allows writing in the other direction.
         :return: true if successful
         """
+        #TODO error handling with timeout
+        await self.mplex_conn.send_message(self.get_flag("CLOSE"), None, self.stream_id)
 
-        if self.local_closed and self.remote_closed:
-            return True
+        remote_lock = ""
+        async with self.stream_lock:
```

Comment:
Can you explain why (in code or PR or both) we have both a stream lock and raw connection lock? I have a bit of trouble following.

Reply:
Got it - in other words, a stream can be modified both by the stream initiator and the stream receiver (e.g. if the receiver closes their corresponding stream and sends a CLOSE), necessitating a lock. Since we have TODOs for this, LGTM.